proofreading section Profiling with MAP (#53)
nesi/perf-training: _lessons/python-scatter/06-profiling_with_MAP.md

---
layout: post
title: Profiling with MAP
permalink: /python-scatter/profiling_MAP
chapter: python-scatter
---

## Objectives

You will:

* learn how to use MAP to profile an MPI code
* learn how to interpret MAP profiling data

## Introduction to profiling parallel codes

Parallel profiling tools provide information about how much time is spent in different parts of your code by different threads/processes.
There are two major ways to collect profiling data: 

>  * *sampling*: data are statistically harvested during run time to determine which parts of the code are most time consuming. This is useful for identiying the hotspots in a code (see [profiling](https://nesi.github.io/perf-training/python-scatter/profiling)).

Reviewer: Relative link again:

```suggestion
 * *sampling*: data are statistically harvested during run time to determine which parts of the code are most time consuming. This is useful for identifying the hotspots in a code (see [profiling](01-profiling.md)).
```

>  * *tracing*: records the activity of instrumented parts of the code in time. This is useful to determine what each thread and process is doing at a particaular point in time. Tracing can reveal load balancing issues or the communication patterns of an MPI program. 

Reviewer 2: thanks for visually highlighting.
Unfortunately it was not clear enough. Tracing measures run time, while sampling "just" statistically analyse call stack snapshots. 
Both have in common that they collect data from all processes ...

Author: I was trying to use a simple sentence to express the usefulness of sampling. I'm not sure the users understand what a call stack is. Suffices to say that sapling provides information about where most of the time is spent in a program...

Reviewer 2: Sorry, that additional comment was just for you. 
The point is, that both methods determine what each thread and processes are doing...

Author: discussed with Mandes and Wolfgang. Decided we would take out this section

MAP can be used both for sampling and tracing.

## Choosing your test case

> Choose your test case carefully, making sure:

Reviewer: ```suggestion
Choose your profiling case carefully, making sure:
```

* it runs in a short time. Due to a possible large overhead from MAP (especially for tracing experiments), the program could run significantly slower than normal.

> * ensure the test case is representative of a production run

Reviewer 2: Do not miss that the run time should not be too short esp. for sampling. Otherwise, we end up with just a few samples and these will not lead to representative analysis.

Author: Thanks @mand35, added "* but the run time should not be too short as this could affect the accuracy of the sampling statistics"

**Note:** keep in mind, that with shortened computation time, the initialisation and finalisation steps may become dominant.

## MAP profiler

On NeSI systems the [ARM MAP](https://www.arm.com/products/development-tools/server-and-hpc/forge/map) profiler is provided as part of the *forge* module (along with the parallel debugger DDT).

MAP is a commercial product, which can profile parallel, multi-threaded and single-threaded C/C++, Fortran, as well as Python code. It can be used without code modification.

> MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without graphical user interface, makes it ieasy to submit job scripts and workflows.

Reviewer: ```suggestion
MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without the graphical user interface, makes it easy to submit job scripts and workflows.
```

> MAP can be launched with a graphical user interface and without. The graphical user interface allows the user to navigate through the code and focus on specific source lines. The "Express Launch", without graphical user interface, makes it ieasy to submit job scripts and workflows.

Reviewer 2: I would first "graphical user interface (GUI)" and then stay with GUI. 
Furthermore, there is a spelling mistake with ieasy. But actually, we want to prevent using such terms. I would say: "The 'Express Launch', without GUI, allows to maintain workflow, since it just requires a modification of the execution line (*srun* command)."

Author: OK thanks.

For more details see the [ARM MAP documentation](https://developer.arm.com/docs/101136/latest/map).

In the following, both the graphical user interface and express launch versions are used with the scatter example.

## Code example

We'll use the *scatter.py* code in directory `mpi` of the *solutions* branch. Start by

```

> git --fetch all

Reviewer: I think you mean:

```suggestion
git fetch --all
```

> git checkout solutions

Reviewer: They would probably need to commit or stash any local changes first but maybe don't need to say that here?

Author: Probably not.... They're likely will not have modified the solutions branch

cd mpi
```

## Using the "Express Launch"

> To use MAP we need to load the *forge* module in our batch script and add `map --profile` in front of the parallel run statements. For example:

Reviewer 2: To be consistent I would also here write "To use MAP we need to load `module load forge` in our batch script ..."

Author: Thanks @mand35. The module load command is written below - no need to duplicate information.

```
module load forge

> map --profile srun --tasks=8 python scatter.py

Reviewer: I don't think the `--tasks=8` is required here? You would set the required number of tasks in an `#SBATCH` line at the top of the submit script and it should be inherited by `srun` - no need to duplicate.

[![example-map-scatter_CPU](images/ARM_MAP_scatter_mpi_CPU.png)](images/ARM_MAP_scatter_mpi_CPU.png)

## Exercises

 * profile the scatter code under the openmp directory using 4 OpenMP threads. Increase the problem size by passing options `-nx 256 -ny 256 -nc 256` to *scatter.py* to make the test run longer.

>  * what is the amount of time spent in computeScatteredWave for the above test case?

Reviewer: ```suggestion
 * what is the amount of time spent in `computeScatteredWave` for the above test case?
```
